src/Workspaces/Core/Portable/CodeRefactorings/FixAllOccurences/DocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,216 @@
+
+            return currentSolution;
+        }
+
+        /// <summary>
+        /// Attempts to apply fix all operations returning, for each updated document, either
+        /// the new syntax root for that document or its new text.  Syntax roots are returned for documents that support
+        /// them, and are used to perform a final cleanup pass for formatting/simplication/etc.  Text is returned for
+        /// documents that don't support syntax.
+        /// </summary>
+        private async Task<Dictionary<DocumentId, (SyntaxNode? node, SourceText? text)>> GetFixedDocumentsAsync(
+            FixAllContext fixAllContext, IProgressTracker progressTracker)
+        {
+            Contract.ThrowIfFalse(fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project
+                or FixAllScope.ContainingMember or FixAllScope.ContainingType);
+
+            var cancellationToken = fixAllContext.CancellationToken;
+
+            using var _1 = progressTracker.ItemCompletedScope();
+            using var _2 = ArrayBuilder<Task<(DocumentId, (SyntaxNode? node, SourceText? text))>>.GetInstance(out var tasks);
+
+            var docIdToNewRootOrText = new Dictionary<DocumentId, (SyntaxNode? node, SourceText? text)>();
+
+            // Process all documents in parallel to get the change for each doc.
+            var documentsAndSpansToFix = await fixAllContext.GetFixAllSpansAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var (document, spans) in documentsAndSpansToFix)
+            {
+                tasks.Add(Task.Run(async () =>
+                {
+                    var newDocument = await this.FixAllAsync(fixAllContext, document, spans).ConfigureAwait(false);
+                    if (newDocument == null || newDocument == document)
+                        return default;
+
+                    // For documents that support syntax, grab the tree so that we can clean it up later.  If it's a
+                    // language that doesn't support that, then just grab the text.
+                    var node = newDocument.SupportsSyntaxTree ? await newDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false) : null;
+                    var text = newDocument.SupportsSyntaxTree ? null : await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                    return (document.Id, (node, text));
+                }, cancellationToken));
+            }
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+
+            foreach (var task in tasks)
+            {
+                var (docId, nodeOrText) = await task.ConfigureAwait(false);
+                if (docId != null)
+                    docIdToNewRootOrText[docId] = nodeOrText;
+            }
+
+            return docIdToNewRootOrText;
+        }
+    }
+}
```

Comment:
the duplicatio nhere is currently a deal breaker for me.  THere are things we can do to unify this though, even without suclassing for codesharing.